File: xeromisc/TrapezoidalProfile.cpp
#include "TrapezoidalProfile.h"
#include <cmath>
#include <iostream>

namespace xero {
    namespace misc {
		TrapezoidalProfile::TrapezoidalProfile() {
			init(0.0, 0.0, 0.0);
		}

        TrapezoidalProfile::TrapezoidalProfile(double max_accel, double max_decel, double max_velocity) {
			init(max_accel, max_decel, max_velocity);
        }

        TrapezoidalProfile::~TrapezoidalProfile() {
        }

		void TrapezoidalProfile::init(double max_accel, double max_decel, double max_velocity) {
			max_accel_ = max_accel;
			max_decel_ = max_decel;
			max_velocity_ = max_velocity;
            isneg_ = false ;
		}

		void TrapezoidalProfile::initFromSettings(SettingsParser &settings, const std::string &prefix) {
			double max_accel = settings.getDouble(prefix + ":max_a");
			double max_decel = settings.getDouble(prefix + ":max_d");
			double max_velocity = settings.getDouble(prefix + ":max_v");
			init(max_accel, max_decel, max_velocity);
		}
		
        double TrapezoidalProfile::intspeed(double t) {
            double ret ;
            if (t < 0.0) {
                ret = start_velocity_ ;
            }
            else if (t < ta_) {
                ret = start_velocity_ + t * max_accel_ ;
            }   
            else if (t < ta_ + tc_) {
                ret = actual_max_velocity_ ;
            }   
            else if (t < ta_ + tc_ + td_) {
				double dt = (t - ta_ - tc_) ;
                ret = actual_max_velocity_ + dt * max_decel_ ;
            }
            else {
                ret = end_velocity_ ;
            }

            return ret ;
        }

        double TrapezoidalProfile::getSpeed(double t) {
            double ret = intspeed(t) ;
            return isneg_ ? -ret : ret ;
        }

        double TrapezoidalProfile::getDistance(double t) {
            double ret ;

            if (t > 0.75)
                ret = 0.0 ;

            if (t < 0.0) {
                ret = 0.0 ;
            }
            else if (t < ta_) {
                ret = start_velocity_ * t + 0.5 * t * t * max_accel_ ;
            }   
            else if (t < ta_ + tc_) {
                ret = start_velocity_ * ta_ + 0.5 * ta_ * ta_ * max_accel_ ;
                ret += (t - ta_) * actual_max_velocity_ ;
            }   
            else if (t < ta_ + tc_ + td_) {
                double dt = t - ta_ - tc_ ;
                ret = start_velocity_ * ta_ + 0.5 * ta_ * ta_ * max_accel_ ;
                ret += tc_ * actual_max_velocity_ ;
                ret += actual_max_velocity_ * dt + 0.5 * dt * dt * max_decel_ ;
            }
            else {
                ret = distance_ ;
            }

            return isneg_ ? -ret : ret ;
        }        

        void TrapezoidalProfile::update(double dist, double start_velocity, double end_velocity) {
            start_velocity_ = std::fabs(start_velocity) ;
            end_velocity_ = std::fabs(end_velocity) ;

            isneg_ = (dist < 0) ;
            distance_ = std::fabs(dist) ;
            ta_ = (max_velocity_ - start_velocity_) / max_accel_ ;
            td_ = (end_velocity_ - max_velocity_) / max_decel_ ;
            double da = start_velocity * ta_ + 0.5 * max_accel_ * ta_ * ta_ ;
            double dd = max_velocity_ * td_ + 0.5 * max_decel_ * td_ * td_ ;
			tc_ = (distance_ - da - dd) / max_velocity_ ;
            type_ = "trapezoid" ;

            if (td_ < 0.0 || da + dd > distance_) {
                //
                // We don't have time to get to the cruising velocity
                //
                double num = (2.0 * distance_ * max_accel_ * max_decel_ + max_decel_ * start_velocity_ * start_velocity_ - max_accel_ * end_velocity_ * end_velocity_) / (max_decel_ - max_accel_) ;
				bool decel_only = false ;
				if (num < 0)
					decel_only = true ;
				else
					actual_max_velocity_ = std::sqrt(num) ;


				if (decel_only || actual_max_velocity_ < start_velocity_) {
                    // 
                    // Just decelerate down to the end
                    //
                    ta_ = 0 ;
                    tc_ = 0 ;
                    td_ = (end_velocity - start_velocity_) / max_decel_ ;
                    actual_max_velocity_ = start_velocity_ ;
                    type_ = "line" ;
                }
                else {
                    //
                    // Can't get to max velocity but can accelerate some
                    // before decelerating
                    //
                    actual_max_velocity_ = std::sqrt(num) ;
                    ta_ = (actual_max_velocity_ -start_velocity_)/ max_accel_ ;
                    td_ = (end_velocity_ -actual_max_velocity_) / max_decel_ ;
                    tc_ = 0 ;
                    type_ = "pyramid" ;
                }
            }
            else {
                //
                // Ok, now figure out the crusing time
                //
                actual_max_velocity_ = max_velocity_ ;                
                tc_ = (distance_ - da - dd) / max_velocity_ ;
            }
        }

        std::string TrapezoidalProfile::toString() {
            std::string ret = "[" + type_ ;
			ret += ", sv " + std::to_string(start_velocity_) ;
			ret += ", mv " + std::to_string(actual_max_velocity_) ;
			ret += ", ev " + std::to_string(end_velocity_) ;
            ret += ", ta " + std::to_string(ta_) ;
            ret += ", tc " + std::to_string(tc_) ;
            ret += ", td " + std::to_string(td_) ;                        
            ret += "]" ;
            return ret ;
        }
    }
}

File: tools/PathViewer/PathViewer/TrapezoidalProfile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PathViewer
{
    class TrapezoidalProfile
    {
        #region private members
        private double m_start_velocity;
        private double m_end_velocity;
        private double m_actual_max_velocity;
        private double m_accel;
        private double m_max_velocity;
        private double m_tc;
        private double m_ta;
        private double m_td;
        private bool m_neg;
        private double m_distance;
        private string m_type;
        #endregion

        public TrapezoidalProfile(double accel, double maxvel)
        {
            m_accel = accel;
            m_max_velocity = maxvel;
        }

        public string TrapezoidalType
        {
            get { return m_type; }
        }

        public double TotalTime
        {
            get { return m_ta + m_tc + m_td; }
        }

        public void Update(double dist, double start, double end)
        {
            m_start_velocity = Math.Abs(start);
            m_end_velocity = Math.Abs(end);

            m_neg = (dist < 0);
            m_distance = Math.Abs(dist);
            m_ta = (m_max_velocity - m_start_velocity) / m_accel;
            m_td = (m_end_velocity - m_max_velocity) / -m_accel;
            double da = m_start_velocity * m_ta + 0.5 * m_accel * m_ta * m_ta;
            double dd = m_max_velocity * m_td + 0.5 * -m_accel * m_td * m_td;
            m_tc = (m_distance - da - dd) / m_max_velocity;
            m_type = "trapezoid";

            if (m_td < 0.0 || da + dd > m_distance)
            {
                //
                // We don't have time to get to the cruising velocity
                //
                double num = (-2.0 * m_distance * m_accel * m_accel + -m_accel * m_start_velocity * m_start_velocity - m_accel * m_end_velocity * m_end_velocity) / (-m_accel * 2);
                if (num < 0 || m_actual_max_velocity < m_start_velocity)
                {
                    // 
                    // Just decelerate down to the end
                    //
                    m_ta = 0;
                    m_tc = 0;
                    m_tc = (m_end_velocity - m_start_velocity) / -m_accel;
                    m_actual_max_velocity = m_start_velocity;
                    m_type = "line";
                }
                else
                {
                    //
                    // Can't get to max velocity but can accelerate some
                    // before decelerating
                    //
                    m_max_velocity = Math.Sqrt(num);
                    m_ta = (m_actual_max_velocity - m_start_velocity) / m_accel;
                    m_td = (m_end_velocity - m_actual_max_velocity) / -m_accel;
                    m_tc = 0;
                    m_type = "pyramid";
                }
            }
            else
            {
                //
                // Ok, now figure out the crusing time
                //
                m_actual_max_velocity = m_max_velocity;
                m_tc = (m_distance - da - dd) / m_max_velocity;
            }
        }

        public double GetAcceleration(double time)
        {
            double ret = 0.0;

            if (time < 0.0)
                ret = m_accel;
            else if (time < m_ta)
                ret = m_accel;
            else if (time < m_ta + m_tc)
                ret = 0.0;
            else if (time < m_ta + m_tc + m_td)
                ret = -m_accel;

            return m_neg ? -ret : ret ;
        }

        public double GetVelocity(double time)
        {
            double ret;
            if (time < 0.0)
            {
                ret = m_start_velocity;
            }
            else if (time < m_ta)
            {
                ret = m_start_velocity + time * m_accel;
            }
            else if (time < m_ta + m_tc)
            {
                ret = m_actual_max_velocity;
            }
            else if (time < m_ta + m_tc + m_td)
            {
                double dt = (time - m_ta - m_tc);
                ret = m_actual_max_velocity - dt * m_accel;
            }
            else
            {
                ret = m_end_velocity;
            }

            return m_neg ? -ret : ret;
        }

        public double GetDistance(double time)
        {
            double ret;

            if (time < 0.0)
            {
                ret = 0.0;
            }
            else if (time < m_ta)
            {
                ret = m_start_velocity * time + 0.5 * time * time * m_accel;
            }
            else if (time < m_ta + m_tc)
            {
                ret = m_start_velocity * m_ta + 0.5 * m_ta * m_ta * m_accel;
                ret += (time - m_ta) * m_actual_max_velocity;
            }
            else if (time < m_ta + m_tc + m_td)
            {
                double dt = time - m_ta - m_tc;
                ret = m_start_velocity * m_ta + 0.5 * m_ta * m_ta * m_accel;
                ret += m_tc * m_actual_max_velocity;
                ret += m_actual_max_velocity * dt - 0.5 * dt * dt * m_accel;
            }
            else
            {
                ret = m_distance;
            }

            return m_neg ? -ret : ret;
        }
    }
}
